src/java/org/apache/cassandra/service/AbstractReadExecutor.java

```diff
@@ -110,16 +116,27 @@ private void makeRequests(ReadCommand readCommand, Iterable<InetAddress> endpoin
             logger.trace("reading {} from {}", readCommand.isDigestQuery() ? "digest" : "data", endpoint);
             if (message == null)
                 message = readCommand.createMessage();
-            MessagingService.instance().sendRRWithFailure(message, endpoint, handler);
+            // Handler adds remote requests latencies to list
+            getMessagingServiceInstance().sendRRWithFailure(message, endpoint, handler);
         }
 
         // We delay the local (potentially blocking) read till the end to avoid stalling remote requests.
         if (hasLocalEndpoint)
         {
+            long localStart = System.nanoTime();
             logger.trace("reading {} locally", readCommand.isDigestQuery() ? "digest" : "data");
             KeyspaceAwareSepQueue.setCurrentKeyspace(command.ksName);
             StageManager.getStage(stage(command)).maybeExecuteImmediately(new LocalReadRunnable(command, handler));
+            synchronized (latencies) {
```

Comment:
To avoid having to manually synchronize access to this list, let's use https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/ConcurrentLinkedQueue.html

Can just use Collection as the top-level type.